lib/vmdb/plugins.rb

```diff
@@ -80,6 +86,48 @@ def register_models
 
     private
 
+    # Determine the version of the specified engine
+    #
+    # If the gem is
+    # - git based, pointing to a branch:   <branch>@<sha>
+    # - git based, pointing to a tag:      <tag>@<sha>
+    # - git based, pointing to a sha:      <sha>
+    # - path based, with git, on a branch: <branch>@<sha>
+    # - path based, with git, on a tag:    <tag>@<sha>
+    # - path based, with git, on a sha:    <sha>
+    # - path based, without git:           nil
+    # - a real gem:                        <gem_version>
+    def version(engine)
+      spec = bundler_specs_by_path[engine.root.to_s]
+
+      case spec.source
+      when Bundler::Source::Git
+        [
+          spec.source.branch || spec.source.options["tag"],
+          spec.source.revision.presence[0, 8]
+        ].compact.join("@").presence
+      when Bundler::Source::Path
+        if engine.root.join(".git").exist?
+          branch = sha = nil
+          Dir.chdir(engine.root) do
+            branch   = `git rev-parse --abbrev-ref HEAD 2>/dev/null`.strip.presence
```

Comment:
This section is for the "path, with git".  In other words, this is for developers who use a `:path` reference in the Gemfile, and that path happens to have a .git directory in it.  In this case (as opposed to the `:git` in a Gemfile case) Bundler does not have any git information and we need to fetch it ourselves.

Reply:
@Fryguy I am aware, but I assume determining the `branch` and `sha` is nothing different than what `Bundler::Source::Git` is doing, and the only reason `Bundler` isn't doing it is because the `:path` flag is telling it that it doesn't need to.

Again, this would require `re-wrapping` the specification, and might not be worth it if it is more code to do and calling into a private API, which might break.  But was curious if you had tried, and you could possibly re-use the code from above.

Reply:
When I have a Bundler::Source::Path it doesn't have this data in it.  I guess I could manufacture a Bundler::Source::Git though, but I don't know what that entails.

Reply:
I tried replacing my code with the Bundler::Source::Git::GitProxy, but it didn't save much code, as I still had to handle removing HEAD and the GitProxy doesn't do the tag lookup.  Bundler::Source::Git works the way it does because the user is specifying *most* of that information in the Gemfile, so it doesn't have to look it up.  It really only needs the GitProxy to find the SHA, which is does the same way I do it https://github.com/bundler/bundler/blob/master/lib/bundler/source/git/git_proxy.rb#L193  (they have --verify, but I don't need it because I'm passing HEAD which will exist)